### DISCUSSION TITLE
How do you mock plugin/module instances?

### DISCUSSION THREAD

**Discussion by aaronbushnell** [markhuot/craft-pest#86] [Q&A] [answered]
If I have a plugin/module that has a service + method like so:
```php
Foo::getInstance()->api->hugeSlowNetworkCall()
```

What's the appropriate way to mock the returned data from this?

I've tried this, but didn't have any luck:
```php
it('parses data correctly', function() {
	mock(Api::class)
		->shouldReceive('hugeSlowNetworkCall')
		->andReturn([ 'foo' => 'bar' ])
		->getMock();

	dd(Foo::getInstance()->api->hugeSlowNetworkCall());
});
```

**Comment by markhuot** [comment, marked as answer]
That's the correct way to mock it. I assume `mock` is coming out of `craft-pest` there?

Something to check would be if inside `Foo::getInstance()` how are they resolving `->api`? It needs to come out of the container to guarantee it gets mocked fully. Otherwise it's up to the order of your tests and making sure that class is not loaded earlier than the mock is set. If you comment out your other tests or throw an `->only()` on this one test does it work then?

**Comment by aaronbushnell** [comment]
Actually, this was my bad @markhuot! Turns out the above _does_ work!

I think the example I was confused about is how to mock a function that exists on the `Foo` plugin class directly. Here's an example:
```php
class Foo extends Plugin
{
  public function init()
  {
    parent::init();
  }

  public function sayHello()
  {
    return 'Hello!';
  }
}
```

```php
it('calls a function from the primary plugin class', function() {
    mock(Foo::class)
        ->shouldReceive([
            'sayHello' => 'Goodbye!',
        ]);

    dd(Foo::getInstance()->sayHello());
});
```

I presume `::getInstance()` is mucking stuff up here. Is there a way to easily mock this type of thing?

**Reply by markhuot** [reply]
Yea, the problem is that plugins init before the tests start so that plugin class can't be mocked because it already exists by the time the test starts.

The short answer is: move any mockable logic off the plugin to a service as they're easier to mock.

The more complicated answer would involve placing some logic inside the plugin class so you could do something like `Foo::getInstance()->fake()` which would cause the instance methods to return fake data instead of live data. But that would all have to be implemented in your plugin code. Pest wouldn't help you with that approach, unfortunately.

**Reply by aaronbushnell** [reply]
This is super helpful, thanks! I'll likely just move the code a bit to make mocking easier. Appreciate it!
